Sources/GRPC/Interceptor/ServerInterceptorPipeline.swift

```diff
@@ -42,56 +42,41 @@ internal final class ServerInterceptorPipeline<Request, Response> {
-  /// the head is first and the tail is last.
+  /// Called when a response part has traversed the interceptor pipeline.
   @usableFromInline
-  internal var _contexts: InterceptorContextList<ServerInterceptorContext<Request, Response>>?
+  internal let _onResponsePart: (GRPCServerResponsePart<Response>, EventLoopPromise<Void>?) -> Void
 
-  /// Returns the next context in the outbound direction for the context at the given index, if one
-  /// exists.
-  /// - Parameter index: The index of the `ServerInterceptorContext` which is requesting the next
-  ///   outbound context.
-  /// - Returns: The `ServerInterceptorContext` or `nil` if one does not exist.
-  @inlinable
-  internal func nextOutboundContext(
-    forIndex index: Int
-  ) -> ServerInterceptorContext<Request, Response>? {
-    return self._context(atIndex: index - 1)
-  }
+  /// Called when a request part has traversed the interceptor pipeline.
+  @usableFromInline
+  internal let _onRequestPart: (GRPCServerRequestPart<Request>) -> Void
 
-  /// Returns the next context in the inbound direction for the context at the given index, if one
-  /// exists.
-  /// - Parameter index: The index of the `ServerInterceptorContext` which is requesting the next
-  ///   inbound context.
-  /// - Returns: The `ServerInterceptorContext` or `nil` if one does not exist.
-  @inlinable
-  internal func nextInboundContext(
-    forIndex index: Int
-  ) -> ServerInterceptorContext<Request, Response>? {
-    return self._context(atIndex: index + 1)
-  }
+  /// The index before the first user interceptor context index. (always -1).
+  @usableFromInline
+  internal let _headIndex: Int
 
-  /// Returns the context for the given index, if one exists.
-  /// - Parameter index: The index of the `ServerInterceptorContext` to return.
-  /// - Returns: The `ServerInterceptorContext` or `nil` if one does not exist for the given index.
-  @inlinable
-  internal func _context(atIndex index: Int) -> ServerInterceptorContext<Request, Response>? {
-    return self._contexts?[checked: index]
-  }
+  /// The index after the last user interceptor context index (i.e. 'userContext.endIndex').
+  @usableFromInline
+  internal let _tailIndex: Int
 
-  /// The context closest to the `NIO.Channel`, i.e. where inbound events originate. This will be
-  /// `nil` once the RPC has completed.
+  /// Contexts for user provided interceptors.
+  @usableFromInline
+  internal var _userContexts: [ServerInterceptorContext<Request, Response>]
+
+  /// Whether the interceptor pipeline is still open. It becomes closed after an 'end' response
+  /// part has traversed the pipeline.
+  @usableFromInline
+  internal var _isOpen = true
+
+  /// The index of the next context on the inbound side of the context at the given index.
   @inlinable
-  internal var head: ServerInterceptorContext<Request, Response>? {
-    return self._contexts?.first
+  internal func _nextInboundIndex(after index: Int) -> Int {
+    return index &+ 1
```

Comment:
Is it worth putting asserts on this unchecked arithmetic? I agree that this _practically speaking_ cannot overflow, but it might be nice to guard against it in debug mode.

Reply:
No harm in it, I added a comment on why it's okay as well.